plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java

```diff
@@ -144,20 +154,118 @@ public void processResponseAsync(
                 responseListener.onResponse(response);
                 return;
             }
-            rewriteResponseDocuments(response, responseListener);
+
+            // if many to one, run rewriteResponseDocuments
+            if (!oneToOne) {
+                rewriteResponseDocuments(response, responseListener);
+            } else {
+                // if one to one, make one hit search response and run rewriteResponseDocuments
+                GroupedActionListener<SearchResponse> combineResponseListener = getCombineResponseGroupedActionListener(
+                    response,
+                    responseListener,
+                    hits
+                );
+                AtomicBoolean isOneHitListenerFailed = new AtomicBoolean(false);
+                ;
+                for (SearchHit hit : hits) {
+                    SearchHit[] newHits = new SearchHit[1];
+                    newHits[0] = hit;
+                    SearchResponse oneHitResponse = SearchResponseUtil.replaceHits(newHits, response);
+                    ActionListener<SearchResponse> oneHitListener = getOneHitListener(combineResponseListener, isOneHitListenerFailed);
+                    rewriteResponseDocuments(oneHitResponse, oneHitListener);
+                    // if any OneHitListener failure, try stop the rest of the predictions
+                    if (isOneHitListenerFailed.get()) {
+                        break;
+                    }
+                }
+            }
+
         } catch (Exception e) {
             if (ignoreFailure) {
                 responseListener.onResponse(response);
             } else {
-                responseListener.onFailure(e);
+                responseListener
+                    .onFailure(new OpenSearchStatusException("Failed to process response: " + e.getMessage(), RestStatus.BAD_REQUEST));
```

Comment:
BAD_REQUEST is a client error: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400 and not the case here. Here we should throw the original exception got from remote server, or at least a 500 error identifying this is a server side error.

Reply:
The most common failure here is client.execute prediction task failed, so it's client errors. And also referencing her is the reason why I am adapting OpenSearchStatusExeception.  

https://github.com/opensearch-project/ml-commons/pull/2801#discussion_r1715957789

Reply:
Http 400 error stands for user side error, meaning that when user get this error, they'll try to figure out the query DSL from their side, the `client` in opensearch is still a server side client, isn't the same one in user side.

Check the exception handling [here](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java#L129) and [here](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java#L186), at least for remote model, you can get the actual exception instead of throwing 400 error.

Reply:
I agree with @zane-neo. We should forward the actual error. Wrapping the client-side error as a server-side error, or vice versa, would be confusing for both the customer and the service team when trying to identify the root cause of any issues.

Reply:
Thanks @dhrubo-os , @zane-neo  and @Zhangxunmt, agree that we should handle different exception properly, and set the default exception is the actual error. 

Added specific handling for OpenSearchStatusException and MLResourceNotFoundException, and the default exception is the actual error in the new commit [here](https://github.com/opensearch-project/ml-commons/pull/2801/commits/6e16254b611f0508c6e64682fbb9f50bdce0a4b8)

Please help review again and approve.